tls/s2n_kem.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "stuffer/s2n_stuffer.h"
+
+#include "tls/s2n_kem.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_mem.h"
+
+int s2n_kem_generate_key_pair(const struct s2n_kem *kem, struct s2n_kem_params *kem_params)
+{
+    notnull_check(kem->generate_keypair);
+    GUARD(s2n_alloc(&kem_params->public_key, kem->publicKeySize));
+    GUARD(s2n_alloc(&kem_params->private_key, kem->privateKeySize));
+    GUARD(kem->generate_keypair(kem_params->public_key.data, kem_params->private_key.data));
+    return 0;
+}
+
+int s2n_kem_generate_shared_secret(const struct s2n_kem *kem, struct s2n_kem_params *kem_params,
```

Comment:
The KEM API kind of bothers me on this one because its doing both, to me encapsulate is just wrapping something I already have. How about `s2n_kem_generate_shared_secret_and_encapsulate`? Or is that too long? I could also be convinced to just do `s2n_encapsulate` since that is what this method is calling on the KEM.

Reply:
Renamed these all to be more consistent after doing more research into standard KEM patterns.